app/services/applied_add_ons/create_service.rb

```diff
@@ -2,74 +2,41 @@
 
 module AppliedAddOns
   class CreateService < BaseService
-    def create(**args)
-      @customer = Customer.find_by(
-        id: args[:customer_id],
-        organization_id: args[:organization_id],
-      )
-
-      @add_on = AddOn.find_by(
-        id: args[:add_on_id],
-        organization_id: args[:organization_id],
-      )
-
-      process_creation(
-        amount_cents: args[:amount_cents] || add_on&.amount_cents,
-        amount_currency: args[:amount_currency] || add_on&.amount_currency,
-      )
+    def self.call(...)
+      new(...).call
     end
 
-    def create_from_api(organization:, args:)
-      @customer = Customer.find_by(
-        external_id: args[:external_customer_id],
-        organization_id: organization.id,
-      )
-
-      @add_on = AddOn.find_by(
-        code: args[:add_on_code],
-        organization_id: organization.id,
-      )
+    def initialize(customer:, add_on:, amount_cents:, amount_currency:)
+      @customer = customer
+      @add_on = add_on
+      @amount_cents = amount_cents
+      @amount_currency = amount_currency
 
-      process_creation(
-        amount_cents: args[:amount_cents] || add_on&.amount_cents,
-        amount_currency: args[:amount_currency] || add_on&.amount_currency,
-      )
+      super
     end
 
-    private
-
-    attr_reader :customer, :add_on
-
-    def check_preconditions
+    def call
       return result.not_found_failure!(resource: 'customer') unless customer
       return result.not_found_failure!(resource: 'add_on') unless add_on
-    end
-
-    def process_creation(amount_cents:, amount_currency:)
-      check_preconditions
-      return result if result.error
 
       applied_add_on = AppliedAddOn.new(
-        customer: customer,
-        add_on: add_on,
-        amount_cents: amount_cents,
-        amount_currency: amount_currency,
+        customer:,
+        add_on:,
+        amount_cents: amount_cents || add_on&.amount_cents,
+        amount_currency: amount_currency || add_on&.amount_currency,
```

Comment:
If add_on is nil, a not_found_failure is already returned, no?

```suggestion
        amount_cents: amount_cents || add_on.amount_cents,
        amount_currency: amount_currency || add_on.amount_currency,
```